%  LaTeX support: latex@mdpi.com
%  In case you need support, please attach all files that are necessary for compiling as well as the log file, and specify the details of your LaTeX setup (which operating system and LaTeX version / tools you are using).

% You need to save the "mdpi.cls" and "mdpi.bst" files into the same folder as this template file.

%=================================================================
\documentclass[electronics,article,submit,moreauthors,pdftex,10pt,a4paper]{mdpi}
\synctex=1
\usepackage{fixme}

\newcommand\floor[1]{\lfloor#1\rfloor}

%--------------------
% Class Options:
%--------------------
% journal
%----------
% Choose between the following MDPI journals:
% actuators, admsci, aerospace, agriculture, agronomy, algorithms, animals, antibiotics, antibodies, antioxidants, applsci, arts, atmosphere, atoms, axioms, batteries, behavsci, beverages, bioengineering, biology, biomedicines, biomimetics, biomolecules, biosensors, brainsci, buildings, carbon, cancers, catalysts, cells, challenges, chemosensors, children, chromatography, climate, coatings, computation, computers, condensedmatter, cosmetics, cryptography, crystals, data, dentistry, designs, diagnostics, diseases, diversity, econometrics, economies, education, electronics, energies, entropy, environments, epigenomes, fermentation, fibers, fishes, fluids, foods, forests, futureinternet, galaxies, games, gels, genealogy, genes, geosciences, geriatrics, healthcare, horticulturae, humanities, hydrology, informatics, information, infrastructures, inorganics, insects, instruments, ijerph, ijfs, ijms, ijgi, inventions, jcdd, jcm, jdb, jfb, jfmk, jimaging, jof, jintelligence, jlpea, jmse, jpm, jrfm, jsan, land, languages, laws, life, literature, lubricants, machines, magnetochemistry, marinedrugs, materials, mathematics, mca, mti, medsci, medicines, membranes, metabolites, metals, microarrays, micromachines, microorganisms, minerals, molbank, molecules, mps, nanomaterials, ncrna, neonatalscreening, nutrients, particles, pathogens, pharmaceuticals, pharmaceutics, pharmacy, philosophies, photonics, plants, polymers, processes, proteomes, publications, recycling, religions, remotesensing, resources, risks, robotics, safety, sensors, separations, sexes, sinusitis, socsci, societies, soils, sports, standards, sustainability, symmetry, systems, technologies, toxics, toxins, universe, urbansci, vaccines, vetsci, viruses, water
%---------
% article
%---------
% The default type of manuscript is article, but can be replaced by:
% addendum, article, book, bookreview, briefreport, casereport, changes, comment, commentary, communication, conceptpaper, correction, conferencereport, expressionofconcern, meetingreport, creative, datadescriptor, discussion, editorial, essay, erratum, hypothesis, interestingimage, letter, newbookreceived, opinion, obituary, projectreport, reply, retraction, review, sciprints, shortnote, supfile, technicalnote
% supfile = supplementary materials
%----------
% submit
%----------
% The class option "submit" will be changed to "accept" by the Editorial Office when the paper is accepted. This will only make changes to the frontpage (e.g. the logo of the journal will get visible), the headings, and the copyright information. Also, line numbering will be removed. Journal info and pagination for accepted papers will also be assigned by the Editorial Office.
%------------------
% moreauthors
%------------------
% If there is only one author the class option oneauthor should be used. Otherwise use the class option moreauthors.
%---------
% pdftex
%---------
% The option pdftex is for use with pdfLaTeX. If eps figure are used, remove the option pdftex and use LaTeX and dvi2pdf.

% *** ACRONYM PACKAGE ***
\usepackage[nolist]{acronym}
\input{acronym} % Name of acronyms file

%=================================================================
\firstpage{1}
\makeatletter
\setcounter{page}{\@firstpage}
\makeatother
\articlenumber{x}
\doinum{10.3390/------}
\pubvolume{xx}
\pubyear{2016}
\copyrightyear{2016}
\externaleditor{Academic Editor: Steven Johnston and Simon J. Cox}
\history{Received: date; Accepted: date; Published: date}
%------------------------------------------------------------------
% The following line should be uncommented if the LaTeX file is uploaded to arXiv.org
%\pdfoutput=1

%=================================================================
% Add packages and commands here. The following packages are loaded in our class file: fontenc, calc, indentfirst, fancyhdr, graphicx, lastpage, ifthen, lineno, float, amsmath, setspace, enumitem, mathpazo, booktabs, titlesec, etoolbox, amsthm, hyphenat, natbib, hyperref, footmisc, geometry, caption, url, mdframed

%=================================================================
%% Please use the following mathematics environments:
 \theoremstyle{mdpi}
 \newcounter{thm}
 \setcounter{thm}{0}
 \newcounter{ex}
 \setcounter{ex}{0}
 \newcounter{re}
 \setcounter{re}{0}

 \newtheorem{Theorem}[thm]{Theorem}
 \newtheorem{Lemma}[thm]{Lemma}
 \newtheorem{Corollary}[thm]{Corollary}
 \newtheorem{Proposition}[thm]{Proposition}

 \theoremstyle{mdpidefinition}
 \newtheorem{Characterization}[thm]{Characterization}
 \newtheorem{Property}[thm]{Property}
 \newtheorem{Problem}[thm]{Problem}
 \newtheorem{Example}[ex]{Example}
 \newtheorem{ExamplesandDefinitions}[ex]{Examples and Definitions}
 \newtheorem{Remark}[re]{Remark}
 \newtheorem{Definition}[thm]{Definition}
%% For proofs, please use the proof environment (the amsthm package is loaded by the MDPI class).

%=================================================================
% Full title of the paper (Capitalized)
\Title{Easy as Pi: A Network Coding Raspberry Pi Testbed}

% Authors, for the paper (add full first names)
\Author{N\'estor J. Hern\'andez Marcano $^{1,2,\dagger}$*, Chres W. S{\o}rensen $^{2,\dagger}$, Juan A. Cabrera G. $^{3}$, Daniel E. Lucani $^{2,\dagger}$ and Frank H. P. Fitzek $^{3}$}
% Authors, for metadata in PDF
\AuthorNames{Nestor Hernandez, Chres Soerensen, Juan Cabrera, Daniel Lucani and Frank Fitzek}

% Affiliations / Addresses (Add [1] after \address if there is only one affiliation.)
\address{%
$^{1}$ \quad Steinwurf ApS; nestor@steinwurf.com\\
$^{2}$ \quad Aalborg University, Department of Electronic Systems; \{nh, cws, del\}@es.aau.dk\\
$^{3}$ \quad Technische Universi\"at Dresden, Deutsche Telekom Chair of Communications and Networks; \{juan.cabrera, frank.fitzek\}@tu-dresden.de}

% Contact information of the corresponding author
\corres{Correspondence: nestor@steinwurf.com, nh@es.aau.dk; Tel.: +45 51 20 03 49}

% Current address and/or shared authorship
\firstnote{Fredrik Bajers Vej 7, Room A3-111. Aalborg, Denmark}

% Simple summary
%\simplesumm{}

% Abstract (Do not use inserted blank lines, i.e. \\)
\abstract{Due to the recurrent nature }

% A single paragraph of about 200 words maximum. For research articles, abstracts should give a pertinent overview of the work. We strongly encourage authors to use the following style of structured abstracts, but without headings: 1) Background: Place the question addressed in a broad context and highlight the purpose of the study; 2) Methods: Describe briefly the main methods or treatments applied; 3) Results: Summarize the article's main findings; and 4) Conclusion: Indicate the main conclusions or interpretations. The abstract should be an objective representation of the article: it must not contain results which are not presented and substantiated in the main text and should not exaggerate the main conclusions.

% Keywords
\keyword{Network Coding; Raspberry Pi; Testbed; C++}

% keyword 1; keyword 2; keyword 3. List three to ten pertinent keywords specific to the article, yet reasonably common within the subject discipline.

% The fields PACS, MSC, and JEL may be left empty or commented out if not applicable
%\PACS{J0101}
%\MSC{}
%\JEL{}

% If this is an expanded version of a conference paper, please cite it here: enter the full citation of your conference paper, and add $^\S$ in the end of the title of this article.
%\conference{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Only for the journal Data:

%\dataset{DOI number or link to the deposited data set in cases where the data set is published or set to be published separately. If the data set is submitted and will be published as a supplement to this paper in the journal Data, this field will be filled by the editors of the journal. In this case, please make sure to submit the data set as a supplement when entering your manuscript into our manuscript editorial system.}

%\datasetlicense{license under which the data set is made available (CC0, CC-BY, CC-BY-SA, CC-BY-NC, etc.)}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Sections that are not mandatory are listed as such. The section titles given are for Articles. Review papers and other article types have a more flexible structure.

%% Only for the journal Gels: Please place the Experimental Section after the Conclusions

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \setcounter{section}{-1} %% Remove this when starting to work on the template.
% \section{How to Use this Template}

% The template details the sections that can be used in a manuscript. Sections that are not mandatory are listed as such. The section titles given are for Articles. Review papers and other article types have a more flexible structure. For any questions, please contact the editorial office of the journal or support@mdpi.com. For LaTeX related questions please contact Janine Daum at latex-support@mdpi.com.

\section{Introduction}

At the beginning, there was darkness and then... bang! \ac{NC}
\cite{ahlswede2000network} appeared to save us from the evilness
of routing.

General introduction. Introduction to topic addressed in the journal.
Review of the State of the Art. Specify why our approach has benefits
and which are they. Indicate contributions.

Our work is organized as follows: Section \ref{sec:testbed} introduces
the testbed system, its setup, connectivity, files configuration and
compilation of the Kodo library. Section \ref{sec:schemes} defines the
coding schemes employed in our study. Section \ref{sec:topologies} some
configurable topologies in the tested. Later, in Section \ref{sec:metrics}
we describe the considered metrics for performance comparison of the
codes deployed in the Raspberry Pi. In Section \ref{sec:results}, we show
the measurements in the Raspberry Pi of the mentioned metrics. Final
conclusions and future work are reviewed in Section \ref{sec:conclusions}.

%The introduction should briefly place the study in a broad context and highlight why it is important. It should define the purpose of the work and its significance. The current state of the research field should be reviewed carefully and key publications should be cited. Please highlight controversial and diverging hypotheses when necessary. Finally, briefly mention the main aim of the work and highlight the main conclusions. As far as possible, please keep the introduction comprehensible to scientists outside your particular field of research.



\section{Testbed Description}
\label{sec:testbed}

We proceed to define the testbed environment by each of its conforming
parts. We later indicate a procedure to set up the testbed. In this procedure,
we summarize all the details regarding on setting the connectivity and
system files for a set of Raspberry Pis in a centralized fashion.
Afterwards, we indicate how to cross-compile the Kodo library in an
easy way. Finally, we provide further information about Kodo itself in terms
of testing, other platforms supported and source code documentation
for further references.

\subsection{System Overview}
\subsection{Installation and Updating}
\subsection{Kodo cross-compilation: From your PC to the Raspberry Pi}

Besides the previous description (\textbf{Include compiling Kodo from the
RasPi from the scratch}), the testbed administrator can compile Kodo in its
personal workstation and transfer the generated binaries directly to
a path in the Raspberry Pi. To achieve this, we get a toolchain that
contains the binaries for the \texttt{raspberry-gxx49-arm-g++} compiler
for the Raspberry Pi. Therefore, we strongly recommend any testbed
administrator to do the following procedure. In what follows, we provide
the instructions considering that the NFS server uses the \texttt{\$HOME}
directory as the working directory. However, the administrator may choose
some other working directory of its preference if desired.

\begin{enumerate}

\item Download the Raspberry Pi toolchain for 64-bit Linux from: \\
\texttt{http://buildbot.steinwurf.dk/toolchains/linux/} to your
\texttt{\$HOME} directory. \\

\item Extract the downloaded file locally in the NFS server. After
this operation, there should be a new directory for the toolchain
in the server. \\

\item Add the \texttt{bin} folder of the toolchain to the \texttt{PATH}
Linux environment variable of the server. This will help the server OS
to recognize the location of the compiler command, which will be needed
later. To do so, edit the \texttt{\$HOME/.profile} to add in a newline:
\texttt{PATH="\$PATH:\$HOME/raspberry-gxx49-arm/bin"}. Save the
\texttt{\$HOME/.profile}. \\

\item Restart the server session in order for the changes made in the
previous step take effect. To verify this, open a new terminal and type:
\texttt{raspberry-gxx49-arm-g++ --version}. A correct binary installation
should return an output similar to:

\texttt{raspberry-gxx49-arm-g++ (crosstool-NG 1.21.0) 4.9.3 20150311 (prerelease)
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.} \\

\item Clone the Kodo repository in the server by executing: \\
\texttt{git clone git://github.com/steinwurf/kodo.git} in \texttt{\$HOME}. \\

\item Navigate to the repository and configure \texttt{waf} by typing:
\texttt{python config.py} and select the 16th ``make specification'' file
for the Raspberry Pi, e.g. option \texttt{[16]cxx\_raspberry\_gxx49\_arm}
presented by the file.

This command configures \texttt{waf} to use the proper compiler and its
required flags to generate the binaries for the Raspberry Pi. If the
configuration was correct, the output will indicate:
\texttt{'configure' finished successfully (X.XXXs)}, where \texttt{X.XXX}
is total time in seconds for configuring the project in the server. \\

\item Execute \texttt{python waf build}. If the build process was
successful, the generated binaries for the Raspberry Pi should be located
in \texttt{build/cxx\_raspberry\_gxx49\_arm} in the Kodo repository.
\textbf{Indicate how the binary files should look like}.

Once this procedure is made, the testbed administrator can relocate the
generated binary files to Raspberry Pi's through the network as desired
by using the \texttt{scp} command during the configuration step.


\end{enumerate}

\subsection{Kodo Builds for the Raspberry Pi, Platform Support and Documentation}

You can check the build status of Kodo, Fifi and other relevant projects through
their respective repository master branch on our buildbot page
\cite{steinwurf2016buildbot}. Our buildbot displays the status of the builds for
Raspbian 8 and GCC 4.9 for the ARM architecture which is the relevant one for the
Raspberry Pi. At the link, you can check build status and build statistics. Also,
documentation about Kodo basics with a tutorial is available at \cite{kododocs}.

\section{Coding Schemes}
\label{sec:schemes}

In this section, we present a description of the considered coding schemes
measured with our the Raspberry Pi testbed. We introduce a definition
for the primitive coding operations, e.g. encoding, decoding and recoding
(where applies) for each coding scheme. Later, we address particular schemes
which are obtained by modifying the basic coding operations.

\subsection{Random Linear Network Coding}

\ac{RLNC} is an example of intra-session \ac{NC}. Here, data symbols from a
single flow are combined with each other. In this type of network coding,
all the original data packets $P_j, j \in [1,g]$, each of $B$ bits, are used
to create coded packets. In the following subsections, we describe the
basic functionalities of \ac{RLNC}. Later, we mention applications that could
potentially benefit from including \ac{RLNC} as a coding scheme.

\subsubsection{Encoding}
With \ac{RLNC}, each coded packet is a random linear
combination of the original set of packets. Hence,
a \ac{l.i.} set of $g$ coded packets, $C_i,\ i \in [1,g]$
is required in order to get the original information.
Each original packet is considered as a concatenation of elements
from a \ac{GF} of a given size $q$, which we denote
$GF(q)$. To create a coded packet, a coding coefficient
$v_{i,j}$, is chosen at random from $GF(q)$ for every
packet $P_j$ and multiplied and added
following the respective \ac{GF} arithmetics. In this way,
a coded packet is:
%
\begin{align} \label{eq:coded_packet}
C_i  &= \bigoplus_{j=1}^{g} v_{i,j} \otimes P_j ,\ \forall i \in [1,g]
\end{align}

To indicate which packets were used to generate a coded packet, one
form is to append its coding coefficients. In this case, the overhead
included for $C_i,\ \forall i \in [1,g]$ by the coding coefficients
is given by:
%
\begin{align} \label{eq:coded_packet_coef}
|v_i| &= \sum_{j=1}^{g} |v_{i,j}| = g \times \lceil \log_{2} (q)
\rceil ~ [\mathrm{bits}]
\end{align}

\subsubsection{Decoding}
To perform decoding, we define $\textbf{C} = \left[C_1 \ldots C_g \right]^T$
and $\textbf{P} = \left[P_1 \ldots P_g \right]^T$. Then, decoding
reduces to solve the linear system $\textbf{C} = \textbf{V} \cdot \textbf{P}$
using Gaussian elimination \cite{fragouli2006network}.
Here, the coding matrix $\textbf{V}$ contains any set of $g$ linearly
independent packets $C_i$ as rows as follows:
%
\begin{align} \label{eq:coding_matrix}
\textbf{V} =
\left[
\begin{array}{c}
        v_1    \\ \hline
        \vdots \\ \hline
        v_g    \\
\end{array}
\right]
=
\left[
\begin{array}{ccc}
        v_{1,1} & \ldots & v_{1,g} \\
        \vdots  & \ddots & \vdots  \\
        v_{g,1} & \ldots & v_{g,g} \\
\end{array}
\right]
\end{align}
%
The decoder begins to compute and remove the contributions from each
of the pivot elements, e.g. leftmost elements in the main diagonal of
\eqref{eq:coding_matrix}, to reduce $\textbf{V}$ to reduced echelon
form. In this way, it is possible to recover the original set of packets.

\subsubsection{Recoding}
Network coding allows intermediate nodes in a network to
recombine (or recode) packets from their sources whether they are
coded or not. In general, a recoded packet should be indistinguishable
from a coded one. Thus, we define a recoded packet as
$R_i$ and its corresponding encoding vector as
$w_i$ with coding coefficients $ \left[w_{i,1} \ldots w_{i,g} \right]$,
as follows:
%
\begin{align}
\label{eq:recoded_packet}
R_i = \bigoplus_{j=1}^{g} w_{i,j} \otimes C_j  ,\ \forall i \in [1,g]
\end{align}

In \eqref{eq:recoded_packet}, $w_{i,j}$ is the coding coefficient that
multiplies $C_j$, uniformly and randomly chosen from $GF(q)$. Any
decoder that collects $R_i, i \in [1,g]$ linearly independent coded packets,
with their respective $w_i$, will be able to decode the data as mentioned
before.

\subsection{Sparse Random Linear Network Coding}

In \ac{SRLNC}, instead of considering all the packets to create a coded
packet like \ac{RLNC} does, here the encoder sets more coding coefficients to
zero when creating a new coded packet.
In practice this can be implemented in
a couple of ways.
%One way to accomplish this is to add a Bernoulli random
%variable, $\text{B}(d)$, as in \ref{eq:sparse_coded_packet}:
The simplest way to accomplish this is to add a Bernoulli random
variable, $\text{B}(d)$, as in \ref{eq:sparse_coded_packet}:

\begin{align} \label{eq:sparse_coded_packet}
    C_i  &= \bigoplus_{j=1}^{g} \text{B}(d)_{j} \, v_{i,j} \otimes P_j ,\ \forall i \in [1,g] ,\
    v_{i,j} > 0,\
    d \in
    \begin{cases}
        {]0,0.5]} & \text{for } q=2 \\
        {]0,1]} & \text{for } q>2
    \end{cases}
\end{align}

We call $d$ the coding density.
Reducing $d$ enables the encoder to decrease the average number of packets that
are mixed into each coded packet. Thereby reducing the complexity of the
encoder that needs to mix less packets, but also on the decoder that in
contrast needs to eliminate less nonzero coding coefficients. The catch of this
approach is that coded packets produced by the encoder becomes more \ac{l.d.}
on each other as the coding complexity is reduced. This incurs in additional
channel overhead.

Generating a coded packet does not imply that it contains any data at all. I.e.
$ \left[ B(d)_{1}, \ldots, B(d)_{g} \right] = \mathbf{0}$.
In that case, one would have to trash the coded packet and generate a new one.

Most coded packets that are produced by the encoder are typically very sparse.
We will therefore rely on a slightly different implementation in KODO
\ref{eq:sparse_coded_packet2} that proved faster for low densities. It is best
expressed as:

\begin{align} \label{eq:sparse_coded_packet2}
    C_i  &= \bigoplus_{j=1}^{\floor{\text{Binom}(g,d)}} v_{i,m} \otimes P_{m} ,\
    %L \subseteq \{1,\dots,g\}: |L|=k ,\
    \forall i \in [1,g] ,\
    v_{i,j} > 0
\end{align}
where
$m = \text{choice}(\{1,\dots,g\}, \text{size}=k, \text{replace}=\text{False})$.
\fxnote{Can this be represented smarter?}


%To solve this, one could trash the coded packet and generate a new coded
%packet. This is not the most efficient because KODO needs to count the
%the number of nonzero coding coefficients in each coded packets to eliminate
%all zero packets.
%
%This is the most correct approach, but the most efficient approach although it is mathematically . 

It is often desired to express the coding density as $d=\frac{k}{g}$. We call
this notation $k-sparse$ where $k$ is the average number of packets combined to
produce a coded packet.

%For low coding density, the encoder might generate a coded packet that consists
%of no coded packets. I.e. the packet consists of all zeroes






%In \ac{SRLNC}, instead of considering all the packets to create a coded
%packet like \ac{RLNC} does, here the encoder sets some coding coefficients to
%zero when creating a new coded packet. This permits to reduce the complexity
%of both the encoding and decoding operations since the encoder (or decoder)
%requires less operations to process the data. Therefore, a new code parameter
%to control the sparsity, the code density $d$, is included to vary the
%number of non-zero packet that are combined. Here are various approaches to
%set the density.
%
%A first method is to specify the number of zeros in a given
%generation size and later generate coded packets with this amount. For example:
%16 zeros in a coding vector of 64 packets where the position of zeros are
%modified at random for every coded packet generated. Another method is to
%define an average number of zeros in the coding vector. Therefore, more or
%less zeros are exactly generated for each coded packet in the process.
%However, if not carefully made, using a sparse coding scheme might increase
%the probability of generating \ac{l.d.} packets which incurs both overhead
%from redundant information and processing delay.

\subsection{Tunable Sparse Network Coding}

\ac{TSNC} can be considered an extension to \ac{SRLNC}. The overall idea is
not only for the encoder to generate sparse coded packets, but also to
choose the sparsity of packets wiser. That is, as the decoder accomulates more
coded packets it becomes more and more likely that the next received
coded packet will be \ac{l.d.}.

The idea of \ac{TSNC} is therefore to gradually increase the coding density
as the decoderes \ac{DOF} increases. This enables TSNC to save a lot of
complexity particulary in the beginning of transmissions, but also to
tailor the coding delay. I.e. approximate the number of coded packets a
decoder needs to receive to decode a generation of $g$ packets. We call
this number the budget, $B$.

Provided a desired budget, $B \geq g$, and the decoders \ac{DOF} it can be
estimated how sparse the coded packets should be to finish a transmission in
time.

In our implementation, it is the encoders responsibility to estimate the coding
density. We use feedback to provide the encoder a better estimate of the
decoders \ac{DOF} at pre-defined points. Namely, when the decoder obtains the
following \ac{DOF}'s:

\begin{equation}\label{eq:feedback_dofs}
    s(k) = g\cdot \frac{2^k -1}{2^k},  \; \text{for } k=0,1,2,\dots,k_t 
\end{equation}
where $k$ is the $k$-th feedback packet that is transmitted during a generation.

The encoder considers the time between feedback packets as regions. It then
estimates a fixed coding density for each region when a feedback packet is
received.

Before the coding density can be estimated it is essential that the encoder
has a good estimate of the decoders \ac{DOF}, but also the remainder of the
total budget. The feedback scheme \ref{eq:feedback_dofs} roughly splits the
remaining \ac{DOF} in halves. This is also the case for the budget that is
split equally among the two new regions. The very last region will be asigned
the full remainder of the total budget.



\section{Topologies}
\label{sec:topologies}
\subsection{Unicast}
\subsection{Broadcast}

\section{Analyzed Metrics}
\label{sec:metrics}
\subsection{Encoding and Decoding Speeds}
\subsection{Latency}
\subsection{Energy}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Results}
\label{sec:results}

This section may be divided by subheadings. It should provide a concise and precise description of the experimental results, their interpretation as well as the experimental conclusions that can be drawn.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \subsection{Subsection}
% \subsubsection{Subsubsection}

% Bulleted lists look like this:
% \begin{itemize}[leftmargin=*,labelsep=4mm]
% \item	First bullet
% \item	Second bullet
% \item	Third bullet
% \end{itemize}

% Numbered lists can be added as follows:
% \begin{enumerate}[leftmargin=*,labelsep=3mm]
% \item	First item
% \item	Second item
% \item	Third item
% \end{enumerate}

% The text continues here.

% \subsection{Figures, Tables and Schemes}

% All figures and tables should be cited in the main text as Figure 1, Table 1, etc.

% \begin{figure}[H]
% \centering
% %\includegraphics[width=3cm]{logo-mdpi}
% \caption{This is a figure, Schemes follow the same formatting. If there are multiple panels, they should be listed as: (\textbf{a}) Description of what is contained in the first panel. (\textbf{b}) Description of what is contained in the second panel. Figures should be placed in the main text near to the first time they are cited. A caption on a single line should be centered.}
% \end{figure}

% \begin{table}[H]
% \caption{This is a table caption. Tables should be placed in the main text near to the first time they are cited.}
% \small % Font size can be changed to match table content. Recommend 10 pt.
% \centering
% \begin{tabular}{ccc}
% \toprule
% \textbf{Title 1}	& \textbf{Title 2}	& \textbf{Title 3}\\
% \midrule
% entry 1		& data			& data\\
% entry 2		& data			& data\\
% \bottomrule
% \end{tabular}
% \end{table}

% \subsection{Formatting of Mathematical Components}

% This is an example of an equation:

% \begin{equation}
% \mathbb{S}
% \end{equation}

%% If the documentclass option "submit" is chosen, please insert a blank line before and after any math environment (equation and eqnarray environments). This ensures correct linenumbering. The blank line should be removed when the documentclass option is changed to "accept" because the text following an equation should not be a new paragraph.
% Please punctuate equations as regular text. Theorem-type environments (including propositions, lemmas, corollaries etc.) can be formatted as follows:
%% Example of a theorem:
% \begin{Theorem}
% Example text of a theorem.
% \end{Theorem}
% The text continues here. Proofs must be formatted as follows:

%% Example of a proof:
% \begin{proof}[Proof of Theorem 1]
% Text of the proof. Note that the phrase `of Theorem 1' is optional if it is clear which theorem is being referred to.
% \end{proof}
% The text continues here.


\section{Conclusion}
\label{sec:conclusions}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\section{Discussion}

%This section may be divided by subheadings. Authors should discuss the results and how they can be interpreted in perspective of previous studies and of the working hypotheses. The findings and their implications should be discussed in the broadest context possible. Future research directions may also be highlighted.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Materials and Methods}

%This section should be divided by subheadings. Materials and Methods should be described with sufficient details to allow others to replicate and build on published results. Please note that publication of your manuscript implicates that you must make all materials, data, and protocols associated with the publication available to readers. Please disclose at the submission stage any restrictions on the availability of materials or information. New methods and protocols should be described in detail while well-established methods can be briefly described and appropriately cited.

%Research manuscripts reporting large datasets that are deposited in a publicly available database should specify where the data have been deposited and provide the relevant accession numbers. If the accession numbers have not yet been obtained at the time of submission, please state that they will be provided during review. They must be provided prior to publication.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\section{Conclusions}

%This section is not mandatory, but can be added to the manuscript if the discussion is unusually long or complex.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\vspace{6pt}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% optional
% \supplementary{The following are available online at www.mdpi.com/link, Figure S1: title, Table S1: title, Video S1: title.}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\acknowledgments{This research has been financed by the CROSSFIRE MITN Marie
Curie project (317126) from the European Comission FP7 framework,
the Green Mobile Cloud project (Grant No. DFF - 0602 - 01372B)
and the TuneSCode project (Grant No. DFF - 1335 - 00125) both granted
by the Danish Council for Independent Research (Det Frie Forskningsr\r{a}d).}



%All sources of funding of the study should be disclosed. Please clearly indicate grants that you have received in support of your research work. Clearly state if you received funds for covering the costs to publish in open access.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \authorcontributions{For research articles with several authors, a short paragraph specifying their individual contributions must be provided. The following statements should be used ``X.X. and Y.Y. conceived and designed the experiments; X.X. performed the experiments; X.X. and Y.Y. analyzed the data; W.W. contributed reagents/materials/analysis tools; Y.Y. wrote the paper.'' Authorship must be limited to those who have contributed substantially to the work reported.}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\conflictofinterests{The authors declare no conflict of interest. The founding sponsors had no role in the design of the study; in the collection, analyses, or interpretation of data; in the writing of the manuscript, and in the decision to publish the results.}

%Declare conflicts of interest or state ``The authors declare no conflict of interest.'' Authors must identify and declare any personal circumstances or interest that may be perceived as inappropriately influencing the representation or interpretation of reported research results. Any role of the funding sponsors in the design of the study; in the collection, analyses or interpretation of data; in the writing of the manuscript, or in the decision to publish the results must be declared in this section. If there is no role, please state ``The founding sponsors had no role in the design of the study; in the collection, analyses, or interpretation of data; in the writing of the manuscript, and in the decision to publish the results''.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% optional
\abbreviations{The following abbreviations are used in this manuscript:\\

\noindent NC: Network Coding\\
RLNC: Random Linear Network Coding\\}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% optional
% \appendix
% \section{}
% The appendix is an optional section that can contain details and data supplemental to the main text. For example, explanations of experimental details that would disrupt the flow of the main text, but nonetheless remain crucial to understanding and reproducing the research shown; figures of replicates for experiments of which representative data is shown in the main text can be added here if brief, or as Supplementary data. Mathemtaical proofs of results not central to the paper can be added as an appendix.

% \section{}
% All appendix sections must be cited in the main text. In the appendixes, Figures, Tables, etc. should be labeled starting with `A', e.g., Figure A1, Figure A2, etc.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\bibliographystyle{mdpi}

%=====================================
% References, variant A: internal bibliography
%=====================================
% \renewcommand\bibname{References}
% \begin{thebibliography}{999}
% % Reference 1
% \bibitem{ref-journal}
% Lastname, F.; Author, T. The title of the cited article. {\em Journal Abbreviation} {\bf 2008}, {\em 10}, 142-149.
% % Reference 2
% \bibitem{ref-book}
% Lastname, F.F.; Author, T. The title of the cited contribution. In {\em The Book Title}; Editor, F., Meditor, A., Eds.; Publishing House: City, Country, 2007; pp. 32-58.
% \end{thebibliography}

%=====================================
% References, variant B: external bibliography
%=====================================
\bibliography{raspi_journal}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% optional
\sampleavailability{The testbed and measurements in this publication are both available from the authors.}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\end{document}

